[WIP] Custom Textures Prefetcher Overhaul Various, Per Extension Count, New Low RAM Calc

ZexaronS opened the pull request:
![gitcontrib_dolphin_customtextures_prefetcheroverhaultake1-perextcount](https://user-images.githubusercontent.com/20824154/27988924-2ce63c9e-642e-11e7-9982-e0fdf8c90c9c.png)

Initially this does not include the option that controls the prefetcher's memory limit to be user configurable in GUI. It could be better for another PR if I don't figure this out soon (1 more day and I need a break), or simply leave it to those who are more GUI experienced. 

But I have already written but not commited some of the code that's in various config files, which should probably make it possible to be at least ini configurable for now, I can update those changes in after I'm sure about which are the only parts needed to make an option ini configurable.

Reopen EDIT: Per-Extension numbers removed in OSD Message, console log only.

goldenercobra commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,74 @@ void HiresTexture::Update()
     }
+
+    // Check FileType And Count - used by prefetcher
+    if ((FileType_Counter_CurrentType == ".dds") || (FileType_Counter_CurrentType == ".DDS"))
```
perhaps use `std::transform(FileType_Counter_CurrentType.begin(), FileType_Counter_CurrentType.end(), FileType_Counter_CurrentType.begin(), ::tolower);` before against checking to cut half the cases? This also catches the cases where the filetype is malformed (ie. `.jPg`).

goldenercobra commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,74 @@ void HiresTexture::Update()
+                                      ".jpg", ".jpe", ".jpeg", ".jfif", ".hdr"};
 
-  std::vector<std::string> filenames =
+  std::vector<std::string> Filepaths_Found =
```
lowercase underscores, `filepaths_found`. This goes for the rest of the variables too.

goldenercobra commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -37,14 +37,24 @@ static std::unordered_map<std::string, std::string> s_textureMap;
 static const std::string s_format_prefix = "tex1_";
 
+// Prefetching counters
+static u32 FileType_Counter_DDS_Amount;
```
I feel like this could be done a bit more elegantly. Perhaps an `unordered_map` with an enum, so all of these variables (with the exception of the last one) can be stored in one object? A struct could also suffice.

riking commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,74 @@ void HiresTexture::Update()
-    SplitPath(rFilename, nullptr, &FileName, nullptr);
+    std::string CurrentFilename;
+    SplitPath(rFilepath, nullptr, &CurrentFilename, nullptr);
+    SplitPath(rFilepath, nullptr, nullptr, &FileType_Counter_CurrentType);
```
if I understand how SplitPath works correctly, you would probably extract both variables in a single call.

riking commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -37,14 +37,24 @@ static std::unordered_map<std::string, std::string> s_textureMap;
+static u32 FileType_Counter_JPG_Amount;  // Common for JPG, JPEG and JFIF variants
+static u32 FileType_Counter_HDR_Amount;
+static u32 FileType_Counter_Total_Amount;
+static std::string FileType_Counter_CurrentType;
```
This can be a local variable.

ZexaronS replied:
But it has to access Update() and Prefetcher() .. no? You mean the CurrentType ?

ZexaronS replied:
Oh you meant the total amount thing

ZexaronS replied:
I think i meant it that it could be used in future in other places, this is still kinda evolving, hence "various", I will put it down in the end if I or others don't come up with ideas that will use it.

riking commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,74 @@ void HiresTexture::Update()
   {
-    std::string FileName;
-    SplitPath(rFilename, nullptr, &FileName, nullptr);
+    std::string CurrentFilename;
```
code style says this should be changed to `std::string filename, extension;` or similar, while you're touching the code.

ZexaronS replied:
Well that would make the rFilename or rFilepath wrong then too no? i think r means root, but i think this should be called the full one, root should be called for the stuff like  /Load/Textures/GZLE01/tex.dds in this case.

JosJuice replied:
> i think r means root

No, it doesn't. `r` prefixes were used in the past to indicate that a variable is a reference.

riking commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -144,21 +196,41 @@ void HiresTexture::Prefetch()
+  size_t absolute_max_mem = Common::MemPhysical();  // Currently "total usable", excl. hw reserved
+  size_t culled_percent_amount = (Common::MemPhysical() * percent_mem_to_cull) / 100;
+  size_t delegated_max_mem = Common::MemPhysical() - culled_percent_amount;
+  size_t below_min_req_mem = 3221225472;  // WIP - 3 GB; 2GB Rec. + 1GB for OS/App/Misc?
```
`1024 * 1024 * 1024 * 3`, and I think there's already constants for this in a header file somewhere.

Also all of these should maybe be `const` or `constexpr`, but that one's a judgement call.

ZexaronS replied:
I used grep and found no mention for these in the repo. Also `const` on `below_min_req_mem` makes a compile error C4307 

> '*' integral constant overflow

.

riking commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -184,34 +256,52 @@ void HiresTexture::Prefetch()
+                      StringFromFormat("%.2f", Pref_sum_total_MB) + " MB in " +
+                      StringFromFormat("%.2f", Pref_total_time_Sec) + " sec, from " + FT_C_DDS_s +
+                      " DDS, " + FT_C_PNG_s + " PNG, " + FT_C_BMP_s + " BMP, " + FT_C_TGA_s +
+                      " TGA, " + FT_C_JPG_s + " JPG, and " + FT_C_HDR_s + " HDR files.",
```
With map of filetype iteration, it would also be nice omit printing of filetypes that have 0 items.

ZexaronS replied:
I haven't figured that out yet, to do it with as less code replication as possible, will do, but I did had it in mind indeed.

goldenercobra replied:
Ternary operator? If x != 0 ? "whatever" + x : ""

goldenercobra commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,74 @@ void HiresTexture::Update()
+    if ((FileType_Counter_CurrentType == ".hdr") || (FileType_Counter_CurrentType == ".HDR"))
+    {
+      FileType_Counter_HDR_Amount++;
+    }
```
Also, all of these should be else if, because if we identify the file we can break from the statement rather than check against all of them.

ZexaronS replied:
Will do that now and see how clang-format does it, but this wouldn't apply if a map is used?

ZexaronS commented:
Now I added the INI support, among other things I had in mind today ... erm tonight, specially the messaging. The next big thing if I can do tomorrow is figuring out how to track the data_size of only the decoding process, this would give the exact and accurate idea of how much memory is being used as overhead compared to on disk raw texture file data size. This is still possible to do but would take more fiddling with a calculation goes like this, pasted from IRC, since I'm a bit exhausted now to retype all of that: 

> here's a way, we get the raw storage size of dds and png without affecting replicating for loops and Level struct, so by calcing you can get DDS,BMP,TGA? mem out of total, which leaves you with the rest that shoud be non-native, and then you calc the storage data sum of all non-native exts and deduce it from the total sum that was stripped of DDS and other native stuff earlier. That's how it would get the amount of overhead i think

-----------------------
Also because of all the recals needed for bytes and MB, I just put them all into double, if you guys are okay with it, only a few actually could be lesser.

The first suggestion of putting per-extension counters into a map would take a bit more time as I'm only beginning to learn C++

-------------------------
-------------------------

![gitcontrib_dolphin_customtextures_prefetcheroverhaultake3-memandcull1](https://user-images.githubusercontent.com/20824154/27991505-6a2b9644-6479-11e7-9409-2e6a8aa4b120.png)


![gitcontrib_dolphin_customtextures_prefetcheroverhaultake2-ct-dxt-descupd1](https://user-images.githubusercontent.com/20824154/27991495-27a6d98c-6479-11e7-90e3-a72f6341c127.png)

![gitcontrib_dolphin_customtextures_prefetcheroverhaultake3-abortwhenbreachlimit](https://user-images.githubusercontent.com/20824154/27991499-344cb706-6479-11e7-9986-8f080c3c60fe.png)

![gitcontrib_dolphin_customtextures_prefetcheroverhaultake2-prefetchdescupd1](https://user-images.githubusercontent.com/20824154/27991497-2f9e6af6-6479-11e7-98eb-4610bb08815b.png)

JosJuice commented on Source/Core/DolphinWX/VideoConfigDiag.cpp:
```diff
@@ -245,10 +245,13 @@ static wxString dump_textures_desc =
-    wxTRANSLATE("Cache custom textures to system RAM on startup.\nThis can require exponentially "
-                "more RAM but fixes possible stuttering.\n\nIf unsure, leave this unchecked.");
+    "Load custom textures from User/Load/Textures/<game_id>/.\nDolphin now supports "
+    "DXTn compresssed textures natively, reducing the amount of required RAM by more than 5 times."
```
Why should this be in a setting description? This isn't a changelog.

ZexaronS replied:
Doesn't it sound like a pretty big change to warrant at least some kind of note, maybe not exactly this one?

JosJuice replied:
It's not especially big compared to some other changes we've had, such as EFB2Tex support for palletted textures.

JosJuice commented on Source/Core/DolphinWX/VideoConfigDiag.cpp:
```diff
@@ -245,10 +245,13 @@ static wxString dump_textures_desc =
+    "\n\nIf unsure, leave this unchecked.");
+static wxString cache_hires_textures_desc = wxTRANSLATE(
+    "Prefetch custom textures to system RAM on startup.\nThis can require exponentially "
+    "more RAM for non-native texture formats such as PNG, but fixes possible stuttering."
```
This requires additional RAM for DDS too (though not as much additional RAM as in the past). Please undo this change.

ZexaronS replied:
That's what you probably heard from others, they are also wrong if the existing counter is correct.

![gitcontrib_dolphin_customtextures_prefetcheroverhaulcomm1-dxt-dds-memevenlower-nooverhead](https://user-images.githubusercontent.com/20824154/27994275-9ea87b9e-64ba-11e7-8fb4-cfc089e44cb9.png)

JosJuice replied:
The settings description is saying that prefetching requires exponentially more RAM *compared to the RAM usage when not prefetching*. It's not compared to the size of files on disk.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -37,14 +37,24 @@ static std::unordered_map<std::string, std::string> s_textureMap;
+static u32 FileType_Counter_JPG_Amount;  // Common for JPG, JPEG and JFIF variants
+static u32 FileType_Counter_HDR_Amount;
+static u32 FileType_Counter_Total_Amount;  // temporairly not local
+static std::string FileType_Counter_CurrentType;
```
This neither follows the style guide nor is consistent with old naming conventions that you shouldn't be using anymore. In variable names, separate words using underscore, and don't use capital letters.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
-      ".jpg"  // Why not? Could be useful for large photo-like textures
-  };
+  std::vector<std::string> extensions{".dds", ".png", ".bmp",  ".tga",  ".tgb",
+                                      ".jpg", ".jpe", ".jpeg", ".jfif", ".hdr"};
```
Why? I thought we were supposed to be heading in the opposite direction?

ZexaronS replied:
This will not be removed, you have no support and did not present any technical reason and/or explanation why such an artificial limit is required or recommended for whatever purpose. This is required for the per-extension console log counts to work at all.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
+                                      ".jpg", ".jpe", ".jpeg", ".jfif", ".hdr"};
 
-  std::vector<std::string> filenames =
+  std::vector<std::string> filepaths_found =
```
`filepaths` as a single word isn't something that I've ever seen be used in Dolphin. Use either `filenames` or `paths`.

ZexaronS replied:
So what if a path doesn't contain the filename ?

spycrab replied:
Then it's still a path.

JosJuice replied:
I don't see how what you're saying makes sense or how it related to my comment.

ZexaronS replied:
Do you meant file_paths then - I'm understanding this as if you want to completely get rid of filepath

I never saw on any PC any text separating filepath or filename in 2 words, so this is some special thing in dolphin then.

JosJuice replied:
What I want is for you to replace `filepaths`. It doesn't matter to me whether the replacement is `file_paths`, `paths`, `filenames`, or something else that's similar.

I've never seen anyone write filepath as one word before (in English, that is).

ZexaronS replied:
Isn't computer english suppose to be a bit different, and be okay with it?

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
 
-    if (FileName.substr(0, code.length()) == code)
+    SplitPath(full_filepath, nullptr, &current_filename, &current_filetype);
+    FileType_Counter_CurrentType = current_filetype;  // seems that it works without &
```
Remove this comment. `FileType_Counter_CurrentType` is an `std::string`, not an `std::string*`, so it makes perfect sense that you don't need `&`.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
+    std::string current_filetype;
+
+    // SplitPath(rFilepath, nullptr, &CurrentFilename, nullptr);
+    // SplitPath(rFilepath, nullptr, nullptr, &FileType_Counter_CurrentType);
```
Don't add commented-out code.

ZexaronS replied:
WIP

Seriously, do you think every time I commit/push im going to extract all these out one-by-one into a temporary file, and then put them back in, like I didn't spend hours on minute things already, I did that for major things, but not one lined things that are sprinkled across.

JosJuice replied:
I seriously expect you to not push things that never would be accepted into Dolphin, unless you have a good reason for it. Whether you do that by moving things to temporary files or by just not putting them in your code at all is up to you.

ZexaronS replied:
I forgot to add WIP to this PR, this was never meant to be left in

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
-    std::string FileName;
-    SplitPath(rFilename, nullptr, &FileName, nullptr);
+    std::string current_filename;
+    std::string current_filetype;
```
`file_type`

ZexaronS replied:
since when is that two words, i've never came across that, what about "base_filename" there's a thousand of those. I don't mind doing it this case but it doesn't make sense to me looking at the overall thing, that's why I'm more keen on making big PRs where a lot of stuff like this can be sweeped up in one go, much less frustrating to code.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
       s_check_new_format = true;
     }
+
+    // Check FileType And Count - used by prefetcher
```
Why is this used by the prefetcher? Isn't it just as relevant when you're not using the prefetcher?

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+  // because this is still WIP - remove_this_when_done
+
+  double prefetch_starttime_total = 0;
+  double prefetch_stoptime_total = 0;
```
Again, underscore before `time`...

ZexaronS replied:
There wasn't one originally, why are you making this as if it's my mistake.

JosJuice replied:
Because:
1. You are using it in a new name.
2. You are using underscore for some word boundaries but not others. (The old names weren't inconsistent in this way.)

ZexaronS replied:
I don't have a problem here it's just that I would prefer in such a case your suggestion should sound in a less scolding way, or maybe I'm making it sound like that, in that case I apologize.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+  double prefetched_size_sum = 0;  // total mem size incl. non-native decompressed textures
+  const double absolute_max_mem = Common::MemPhysical();
+  // Currently "total usable", excl. HW reserved
+  const double culled_percent_amount = (Common::MemPhysical() * percent_mem_to_cull) / 100;
```
"percent amount" doesn't make much sense IMO. Either it's a percentage or an amount.

ZexaronS replied:
Agreed. Don't need the full description, it should be obvious that it's derived from the percent.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+  double prefetch_stoptime_total = 0;
+
+  // Default 30% -- Todo GUI User Configurable, INI Only for now
+  double percent_mem_to_cull = 30.0;  // aka culled_percent
```
Why do you need to call this by two names?

ZexaronS replied:
The aka is the old name, I tried the new one since culled_percent sounds more specific, and it was very similar to culled_percent_amount, which now is going to change to culled_amount

ZexaronS replied:
Secondairly, I used the culled_percent in the GFX config since it's shorter, there were many variations before this was pushed, you only see what I eventually come up with, but it doesn't mean it's already set in stone.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+    {
+      // Successful Manual Override - Dont report values in OSD, routine stats notification will do
+      OSD::AddMessage("Prefetcher: Maximum delegated memory limit has been manually overriden "
+                      "by user INI settings.");
```
Please don't use OSD messages unless they really are necessary. They can be annoying and are also not a good way to convey information (due to them disappearing automatically, not scaling with the resolution, and not being translated).

ZexaronS replied:
Doesn't that make them totally useless then? What about, we make them sacalabe, making them translated, and making them not disappear so quick?  Sometimes it better to make other things adapt to you that you adapting to them. 
But hey I forgot to add the timeout there.

Tilka replied:
Yes, it does make them quite useless and we're aiming to reduce them rather than add more.

JosJuice replied:
Yes, it would be nice to make them scaleable and translateable. Would you like to work on that?

They're still going to be annoying, though. Especially if you make them not disappear so quick.

ZexaronS replied:
I seriously thinking about tackling OSD Messages to get my revenge hah
But ... after a break, still didn't had one since VS2017 install (which was really a 1 week mega pc maintenance operation)

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+                " is below the highest bound. Resetting to highest allowable value.",
+            20000);
+        WARN_LOG(VIDEO, "Prefetcher: Mem limit cull percent INI override out of "
+                        "higher bound. Setting to highest");
```
"out of higher bound" sounds ungrammatical to me

spycrab replied:
I'd use something among those lines:
> Mem(ory) limit cull(ing?) percentage exceeds the maximum of UPPER_BOUND %. Lowering the percentage to the upper bound.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+  if (Config::Get(Config::GFX_HIRESTEX_OVERRIDE_PREFETCH_CULLED_PERCENT) != 30.0)
+  {
+    if ((Config::Get(Config::GFX_HIRESTEX_OVERRIDE_PREFETCH_CULLED_PERCENT) < 5.0) ||
+        (Config::Get(Config::GFX_HIRESTEX_OVERRIDE_PREFETCH_CULLED_PERCENT) > 90.0))
```
You can skip this `if`, since you check `< 5.0` and `> 90.0` again below.

ZexaronS replied:
But the OR is needed to fire the same message in the Else, otherwise the Else message would have to be duplicated no?

JosJuice replied:
No. Just put the `else` after the `else if (percent > 90.0)`.

ZexaronS replied:
Okay, but I need to wrap my head around it ... to learn something along the way, would be bad if I didn't get the most out of this run.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+      }
+      else if (Config::Get(Config::GFX_HIRESTEX_OVERRIDE_PREFETCH_CULLED_PERCENT) > 90.0)
+      {
+        percent_mem_to_cull = 90.0;
```
Please make constants for 90.0, 30.0 and 5.0.

ZexaronS replied:
I don't understand.

JosJuice replied:
Assign 90.0, 30.0 and 5.0 to `constexpr` variables, then use those variables in the code instead of writing 90.0, 30.0 and 5.0 everywhere. That way, if someone wants to change those values in the future, they only have to do it in one place instead of several places. The code will also be easier to understand.

spycrab replied:
Also see [this indepth explanation on why "magic numbers" are bad](https://blogs.msdn.microsoft.com/alfredth/2012/02/14/magic-numbers-are-bad/).

ZexaronS replied:
However, the vars that contain `Common::MemPhysical()` can't seem to be `constexpr` only `const` , others are ok.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+  {
+    OSD::AddMessage("Prefetcher: Warning! System RAM of " +
+                        StringFromFormat("%.2f MB", absolute_max_mem / (1024.0 * 1024.0)) +
+                        " is below Dolphin's minimum recommended specifications.",
```
It doesn't make much sense to say this for less than 3 GB when the recommended specifications listed on the website FAQ says 2 GB. If you meant that it's not fine for prefetching, say that it's not fine for prefetching, not that it isn't fine for Dolphin in general.

ZexaronS replied:
WIP, I don't what context that 2GB on Dolphin's FAQ is using. 2GB total installed, 2GB of free RAM (which means `installed - used = atleast 2GB free left`)

JosJuice replied:
It's about hardware requirements, so it's 2 GB total installed.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
-      (sys_mem / 2 < recommended_min_mem) ? (sys_mem / 2) : (sys_mem - recommended_min_mem);
-  u32 starttime = Common::Timer::GetTimeMs();
-  for (const auto& entry : s_textureMap)
+  // Routine Stat Notification
```
Uh, what? What is a "routine stat notification" and why are you adding one?

ZexaronS replied:
Didn't found better wording yesterday, but left in there as a WIP reminder; could find them today.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
   {
-    const std::string& base_filename = entry.first;
+    const std::string& base_filename = texmap_ref.first;
+    // const std::string& full_filepath = texmap_ref.second;
```
Again, don't add commented-out code.

ZexaronS replied:
WIP

JosJuice replied:
If you don't want reviews on a PR for the time being, you should say so in the PR description.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
+    // with the right formula that would return the amount of overhead
+    // (for 14 Megs of PNGs mem is ~350MB)
+    // to get the per-extension time probably this whole for loop would have to be redone
+    // I was not able to do this with full day hours while leaving "m_levels" unchanged
```
Please don't put rambling like this in Dolphin's source code.

ZexaronS replied:
That's just WIP, which means it's for the duration of this PR evolvement. It's also a good way to remind my self and the others of the plans, gives an option to discuss here if anyone has any better ideas. One of them suggested yesterday that line in HiresTextures.cpp ` u8* data = SOIL_load_image_from_memory(buffer.data(), static_cast<int>(buffer.size()), &width,
                                         &height, &channels, SOIL_LOAD_RGBA);`

Tried a bit only, but I will more deeply try to do something today, if lucky, might even get overhead calculation in by the end of the day.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -135,30 +188,111 @@ void HiresTexture::Update()
         // try to get this mutex first, so the video thread is allow to get the real mutex faster
+        // JonnyH: so the prefetch() is running in a tight loop, and it's using that to allow
+        // "at least" 1 Search() run per cycle
+        // merry: exactly!
```
Pasting what seems like IRC logs isn't a good way to clarify something.

ZexaronS replied:
I asked the IRC what to do about this, they provided none, which to me from now on should be equal to "not bad enough for them to care stopping the pressed and replying" so I had to come up with a solution my self. This is the philosophy that I'm using from now on, since I am seeing that being used by other's as well, pretty much all the other config names are subjectively decided, what about all the past things that I have to fix now, or I'm have to going to sit in IRC for hours to get things pre-approved, already I tried that, it didn't work so good.

Whining and drama beside, I would just rather skip and hear the suggestions here, either was the point has to be made across so that any future confusion and required-time-to-brainstorm-before-understanding for newcomers and old cats is made obsolete.

JosJuice replied:
> I asked the IRC what to do about this, they provided none, which to me from now on should be equal to "not bad enough for them to care stopping the pressed and replying"

It's not equal to that.

I think many people on IRC are choosing to not respond to you because discussions with you tend to be a waste of time. That doesn't mean that we don't care about not making the source code worse.

> so I had to come up with a solution my self.

A solution for what? It wasn't broken, and I can't see anything you changed other than adding this comment.

> or I'm have to going to sit in IRC for hours to get things pre-approved

I don't recommend sitting in IRC for hours to get things approved. What I recommend is not making any PRs if they quality of them is going to be as bad as this one. We've already spent countless of hours on helping you on IRC, and most of us probably aren't willing to put much more time into helping you with every detail of every PR you're making.

ZexaronS replied:
After I talked about some unrelated thing about the mutex locks, Someone else (JonnyH i think) gone through and wasted 20 minutes on trying to figure out how these mutex locks 

I think this is a big enough "detail" to have it properly commented so not me or anyone else wastes 20 or more minutes figuring it out. Don't you think that is frustrating, people like me who are eager to do improvements and write code, I should be trying to avoid spending time looking at the code, it's all connected to the part of not being paid, every now and then it doesn't hurt to put a bit more effort in than necessary, I would not be even bringing that up, but across the board, if every piece of code takes so long to figure out it crosses a line, then it becomes a problem. I'm prepared to do more than necessary, but the requirement is it has to be as enjoyable as possible.

JosJuice replied:
Sure, properly commenting it is a good idea. But what you've added isn't a proper comment.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -184,34 +318,56 @@ void HiresTexture::Prefetch()
     }
 
-    if (size_sum > max_mem)
+    // The important part - Stops prefetching to avoid getting out of RAM completely, crash prone
```
I thought the important part of prefetching textures was to prefetch textures...

ZexaronS replied:
Wasn't it doing that before. That statement is suppose to be relative to the low ram measures from before. Maybe that relativity could be made more obvious.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -184,34 +318,56 @@ void HiresTexture::Prefetch()
       if (iter != s_textureCache.end())
       {
-        for (const Level& l : iter->second->m_levels)
+        for (const Level& tex : iter->second->m_levels)
```
If you're going to change this variable name, please change it to a full word. (Preferably `level`, since that matches `m_levels` and probably is what `l` was short for.)

ZexaronS replied:
For what will be one more word that is so similar to a global one, to make things even more confusing? I wasn't able to figure out, evne with help from IRC, what that actually does in technical detail so I has no proper idea for that name. On the other hand we know that this counts all the memory of all textures that are prefetched, that could give some ideas how to name this, the term level simply doesn't mean anything for this local case, or if it does, then I'm open to hearing an explanation.

JosJuice replied:
It's iterating through the different mipmap levels of each texture.

ZexaronS replied:
Actually I asked about that and they told me that the `if (base_filename.find("_mip") == std::string::npos)` line in which the data size iterator is sitting it, means everything else except the _mip files. I'm not sure yet what it does if there is a _mip, if it even counts those.

Otherwise this wouldn't work, since there's no _mip files in the Hypatia's Texture Pack for Zelda WW.

JosJuice replied:
> Actually I asked about that and they told me that the `if (base_filename.find("_mip") == std::string::npos)` line in which the data size iterator is sitting it, means everything else except the _mip files. I'm not sure yet what it does if there is a _mip, if it even counts those.

The `Load` function that's called here will load the mipmap files: https://github.com/ZexaronS/dolphin/blob/f82380b7107195983909592168b0ace3e237e05d/Source/Core/VideoCommon/HiresTextures.cpp#L311

> Otherwise this wouldn't work, since there's no _mip files in the Hypatia's Texture Pack for Zelda WW.

A texture that isn't using mipmaps is considered by Dolphin to have exactly one mipmap level.

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -184,34 +318,56 @@ void HiresTexture::Prefetch()
+  const double Pref_sum_total_MB = prefetched_size_sum / (1024.0 * 1024.0);
+  const double Pref_total_time_Sec = (prefetch_stoptime_total - prefetch_starttime_total) / 1000.0;
+
+  OSD::AddMessage("Prefetcher: Total of " + FT_C_Total_s + " custom textures loaded, " +
```
Again, please don't use OSD messages unless it's really necessary. This is applicable for most of the OSD messages that you've added in this PR, but I especially don't think that the user should have to see a count like this every time they start a game using custom textures.

ZexaronS replied:
Unfortunately this is a major feature of this PR  for me. Without this the PR is worthless. The whole point are more OSD messages for details that I alwa

Remember, as the -dev channel people said, some of them don't really care about custom textures and other things what I suggested, but they don't mind if I go and do things myself, so if everyone else puts in what they want, what can't I.

Since when is opensource some subjectivelly controlled thing ... I'm not paid to babysit the community.

leoetlino replied:
> if everyone else puts in what they want, what can't I.

You can. It will just stay in your fork.

JosJuice replied:
> Unfortunately this is a major feature of this PR for me. Without this the PR is worthless. The whole point are more OSD messages for details that I alwa

So this PR is worthless in my opinion, then.
> if everyone else puts in what they want, what can't I.

We won't stop you from adding what you want to add to your own copies of Dolphin, but that doesn't mean that we're going to accept any and all changes into upstream. If we did that, Dolphin would be an awful mess of half-broken features and terrible UX. (I'm not saying that the implementation of this PR is broken, but I do think it makes the UX worse.)

> I'm not paid to babysit the community.

And we're not paid to babysit you.

ZexaronS replied:
So 2 hours since I woke up, looking back now, I was more frustrated with the comments and the scolding-sounding of the suggestsions, than the actual content of the suggestions themselfs.  

But it's my fault because I forgot to add WIP moniker.

Tilka commented:
I believe the original intend of this PR was to allow users to configure a maximum amount of memory to be used for prefetched textures, so it would be nice if you could focus on that. We are definitely not going to add more OSD messages.

spycrab commented:
@dolphin-emu-bot rebuild

ZexaronS commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -91,32 +101,75 @@ void HiresTexture::Update()
-      ".jpg"  // Why not? Could be useful for large photo-like textures
-  };
+  std::vector<std::string> extensions{".dds", ".png", ".bmp",  ".tga",  ".tgb",
+                                      ".jpg", ".jpe", ".jpeg", ".jfif", ".hdr"};
```
This is what SOIL supports, why limit it.

JosJuice replied:
We want to drop technical debt (meaning we want to drop SOIL), not keep it just because it might be useful for someone. Otherwise we would still be having things like a Direct3D 9 backend...

ZexaronS replied:
I asked exactly that on IRC, will we get BC7 support or not, will SOIL be dropped or not, there was nothing conclusive and some people were even cracking jokes or it looked like it, I don't understand english jokes very well.

JosJuice replied:
I thought we were clear enough about wanting to drop SOIL. (But I could be misremembering, considering how many things were said during those hours...)

ZexaronS replied:
Timeframe or ETA until SOIL dropped ?

JosJuice replied:
That's hard to know.

ZexaronS commented:
So one person can make a subjective decision for the whole project ?

JosJuice commented:
> So one person can make a subjective decision for the whole project ?

They can, if nobody else has any objections. Right now, you (a single person) are trying to add a feature that both I and @Tilka are objecting to.

spycrab commented:
> So one person can make a subjective decision for the whole project ?

Nope and most of the things @JosJuice pointed out are rather obvious and could've been avoided by reading ``Contributing.md`` and cleaning up your PR beforehand.

> if everyone else puts in what they want, what can't I.

We don't allow that.

ZexaronS commented:
But that's still a too short time period, plus, @JosJuice you said you don't even focus on this area, I saw that many times.

When I had my comments about the GUI, the people responsible for that were serious about getting their stuff across and denying my input. So I won't meddle in the GUI, I found my place here, but now I can't even do anything here.

spycrab commented:
> When I had my comments about the GUI, the people responsible for that were serious about getting their stuff across and denying my input. So I won't meddle in the GUI, I found my place here, but now I can't even do anything here.

Input / Criticism is appreciated when it's:
* clearly put
* coherent
* on topic
* constructive (e.g. saying what you'd like to be done instead, why you think something is bad / confusing etc.)

There are a lot of UI related arguments, but they are *mostly* constructive due to people adhering to the principles listed above.

If it isn't however then people are just going to disregard it (the more friendly option) or ignore it.

*Anyway, done sticking my hand in to a beehive here, I'm out*

Tilka commented:
> So one person can make a subjective decision for the whole project without much of debate going on ?

The issues of OSD messages have been brought up and agreed upon several times in the past.

JosJuice commented:
> @JosJuice you said you don't even focus on this area, I saw that many times.

That's right. Because of that, I'm not able to give much input on things that would be specific to the custom texture code. I can however give input on things that would be a bad idea anywhere in Dolphin - that's what I've been doing when reviewing this PR.

> When I had my comments about the GUI, the people responsible for that were serious about getting their stuff across and denying my input. So I won't meddle in the GUI, I found my place here, but now I can't even do anything here.

The reason we don't want to accept your changes is because most of your changes are bad. Trying to argue about it isn't going to make your changes better.

> Can I please have the offending OSD messages exactly labeled, which ones, would like to quickly finish this up today.

These ones are not acceptable at all in my opinion:

https://github.com/ZexaronS/dolphin/blob/f82380b7107195983909592168b0ace3e237e05d/Source/Core/VideoCommon/HiresTextures.cpp#L242
https://github.com/ZexaronS/dolphin/blob/f82380b7107195983909592168b0ace3e237e05d/Source/Core/VideoCommon/HiresTextures.cpp#L270

And this one shouldn't contain as much detail as it does (i.e. you should drop the per-filetype reporting):

https://github.com/ZexaronS/dolphin/blob/f82380b7107195983909592168b0ace3e237e05d/Source/Core/VideoCommon/HiresTextures.cpp#L270

I can't say whether the other ones are okay or not.

ZexaronS commented:
Ah, so the OSD Messages also fall into the GUI category, now it makes all sense, what was I thinking. 

Whatever, I can probably absorb all this as experience, start is always the hardest, but it's still so hard to delete a chunk I spent fiddling on, even tho it's my own problem, maybe if some kind of compromise could be done. I agree these messages are not that important, they're not even important for logging, but at least the warn/info logs could be left in ?

@JosJuice But how "Anywhere", this is a deep setting, none of these is shown unless the user ticks "load custom textures" AND "prefetch custom textures" ... that's my biggest argument, it's not some mainstream/usual area. I thought I found a pretty deep hole to relax, I guess not.

JosJuice commented:
> @JosJuice But how "Anywhere", this is a deep setting, none of these is shown unless the user ticks "load custom textures" AND "prefetch custom textures" ... that's my biggest argument, it's not some mainstream/usual area. I thought I found a pretty deep hole to relax, I guess not.

It's true that this only will affect some users, but prefetching isn't exactly an obscure feature, so that doesn't make it okay.

ZexaronS commented:
It is more of an obscure feature now that DXT(DDS) native support and GPU decoding make things a lot better, imo. Yes I did test that myself some time ago.

- I'm okay with removing the "has been manually overriden" OSD
- I'm not so keen about removing the //routine status notification, but oh well.
- But I'm definitely not removing per-extension counting, I can improve it as suggested to not show the ones which are zero. Infact I was planing to expand so it would show the amount of overhead for non-native textures.

Those are my final decisions. 

Now I spend 2 days thinkering about this, and you guys, have only gone through a few minutes of the thought process, that in it selfs feels unbalanced.

Tilka commented:
Well, you better find someone else to merge this PR then.

ZexaronS commented:
I think I should have put the WIP moniker up, I thought about it but forgot, sorry about that.

Just one more thought from before: I understand there's problems with OSD Messages, now that I'm aware, but this is a technical thing, these obscure OSD messages probably aren't annoying any of you that personally right? Why don't we fix the rascals then, instead of putting it down on me saying it's my fault I even thought of using them, I feel like a little kid being scolded.

That said, better get back to work or all this would be for nothing.

JosJuice commented:
> these obscure OSD messages probably aren't annoying any of you that personally right?

They are.

> Why don't we fix the rascals then, instead of putting it down on me saying it's my fault I even thought of using them, I feel like a little kid being scolded.

The biggest problem with this PR isn't that it uses OSD messages, it's that what it does (other than adding a configurable memory limit) isn't especially useful to begin with. (This is why it's a good idea to cleanly separate your changes, by the way. If we would've wanted to merge the configurable memory limits but not the per-filetype counting, we wouldn't be able to do that as long as it's all in one PR. Not to mention that cleanly separating the changes makes them easier to review.)

And sure, I would like to see OSD messages be improved, but nobody seems to want to work on that right now. (Unless you want to?) But even if we do that, OSD messages are still going to be annoying and are still going to have the problem of disappearing quicker than you might be able to read them. (Some people read very fast and some people move very slow, so no fixed timeout value that we select can make everyone happy.) So I would be against adding OSD messages like these even if OSD messages show up at appropriate sizes and are translatable.

> That said, better get back to work or all this would be for nothing.

Yes, please get back to work with fixing our comments if you want to get anything merged.

ZexaronS commented:
> They are.

> So I would be against adding OSD messages like these even if OSD messages show up at appropriate sizes and are translatable.

Ladies and gentlement, why are OSD Messages still included in dolphin ?

JosJuice commented:
A small amount of the OSD messages are useful enough to actually warrant having them.

ZexaronS commented:
Ah but only those that haven't show any useful information. You didn't know there's more than 1000 PNGs in what is suppose to be a DDS custom textures pack. ... I thought so as well, the the fixed message should fix this confusion from ever happening again for everyone, how the heck is this BAD ?

JosJuice commented:
Users rarely care about the file types in a texture pack. (Even in the case that you brought up, me talking about Hypatia's WW pack on IRC, knowing that there were PNGs in the pack wasn't useful.) In the rare case that users do care, I don't see why they can't check it using another program.

Counting texture formats isn't a completely pointless feature to have in Dolphin, but I don't think it's useful enough to have considering the impact on the code.

ZexaronS commented:
Maybe you're going with the _"developers who devote their free time to an opensource project are bound by what the users want and need"_ philosophy a bit too strong, no?

JosJuice commented:
How is caring about end-users a bad thing?

ZexaronS commented:
It's not, but, I think with the existing OSD Message, that I'm merely improving, it's a bit too far to use that philosophy to block it completely, while having in mind that while I didn't particularly got the news in a good mood, I got over it and conceded to those, I can't call them suggestions anymore, but demands. 

I agreed earlier that showing 0 JPG, 0 BMP ..etc, isn't particularly enjoyable, but I haven't figured that out yet, so I will make anything to make the primary per-ext thing as good and as less annoying as possible.

In the meantime, check out how you like the latest changes. In a minute.

spycrab commented:
@dolphin-emu-bot rebuild

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -144,22 +190,81 @@ void HiresTexture::Prefetch()
       {
-        // try to get this mutex first, so the video thread is allow to get the real mutex faster
+        // prefetch() is running in a tight loop, and it's using that to allow
+        // "at least" 1 Search() run per cycle
```
Why is "at least" in quotation marks?

JosJuice commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -144,22 +190,81 @@ void HiresTexture::Prefetch()
+      {
+        percent_mem_to_cull = 5.0;
+        OSD::AddMessage(
+            "Prefetcher: User INI setting for 'culled cercent' of " +
```
`cercent` -> `percent`

ZexaronS commented on Source/Core/VideoCommon/HiresTextures.cpp:
```diff
@@ -144,22 +190,81 @@ void HiresTexture::Prefetch()
     {
       {
-        // try to get this mutex first, so the video thread is allow to get the real mutex faster
+        // prefetch() is running in a tight loop, and it's using that to allow
```
That's because JonnyH was talking about mutexes when figuring things out - let's take a look again if the subject was changed or something else.

![gitcontrib_dolphin_customtextures_prefetcheroverhaulcomm2_mutexlocksirclogjonnyh](https://user-images.githubusercontent.com/20824154/27996231-21e049ba-64de-11e7-80c3-8226e630a93e.png)

JosJuice replied:
Yes, but anyone who's reading the code isn't going to have the context of that IRC log. You'll have to rewrite it so that it makes sense.

ZexaronS replied:
I don't know myself how does that work with mutex locks sorry

JosJuice replied:
Then you should leave the comment the way it was.

stenzek commented:
-1 from me also for showing this information. I don't think a breakdown of file formats is particularly relevant to users texture packs. Probably the only important OSD message here would be if prefetching failed for some reason.

There's also other errors that can occur when loading, for example, incorrect number of mipmaps, or incorrect dimensions, neither of which are logged to the OSD (which is appropriate, given it's only applicable to texture pack authors, who can turn on logging). Perhaps another logging category would be worthwhile here though, to separate it from the noise of the video backend logging.

I've stated this before, but in my opinion, this sort of functionality would be best suited for a "Texture Pack Tool", either integrated into, or separate from Dolphin. This tool could show a break down of textures, memory usage, and make any errors clearly visible. Such a tool could also be used to convert texture packs to compressed formats that are appropriate for their GPU (remember, not all GPUs support DXT1-5), as real-time transcoding isn't really an option, especially for the more complex compressed formats.

ZexaronS commented:
given it's only applicable to texture pack authors -> subjective opinion

ZexaronS commented:
I don't really understand why it has to be relevant to other users who aren't even part of any discussions, have no interest participating.

ZexaronS commented:
if (change_i_don_like_) { return insert_our_users_dont_need_it_excuse; } 

I was even thinking of helping with the OSD Messages, I don't know if I'll ever get back into the mood for that.

ZexaronS commented:
Reopen Update:
Removed new OSD messages and replacing them with log messages.
Removed per-extension counts in main OSD Message

Not buildable at this time, some LINK issues.

ZexaronS commented:
WIP Label can be removed now, this is as far as this PR was meant to go, the ability for the user to limit how much RAM is used for custom texture prefetching is made with the override value which is ini only, I will avoid getting into the middle while the transition to QT GUI is being made.

All the changes in requested changes label were fixed as well. 

I will report once I fix the LNK2001 issue.

JosJuice commented:
> WIP Label can be removed now

You can remove the WIP label from the title on your own when you feel that that's appropriate.

> All the changes in requested changes label were fixed as well.

You haven't undone the addition of more supported filename extensions, or the mention of non-native texture formats in the DolphinWX setting description. I've explained before why those changes aren't appropriate.

There are also still questionable variable names – please read Contributing.md and follow the naming style that we use.

ZexaronS commented:
@JosJuice The support for more filename extension was never part of the compromise, only the inclusion of per-extension counts into the OSD messages was. The per-extension counts remain in console log output only, the people who were present in the IRC Chat at my last mention before I was banned agreed to the removal of the per-extension in the OSD, there was no problem with the whole custom textures feature supporting, except with you.

I have said many times this PR is pointless if that get's dropped, you're asking me to drop the main feature of the PR, that means you're asking me to remove every modification for this PR to get passed, so I will merge thin air, you are also, so far, the only person with such a ridicolous demands, everyone else was specifically against the inclusion of per-extension counts in the OSD message.

My stand is that what SOIL supports then there should be no aritifical limit on top of that, you have presented no technical reasons and no explanations as to why such an artificial limit is required. Remove your blockade to stop this debacle from going even deeper.

JosJuice commented:
Like I've said before, there is a rough agreement between several people that we want to stop using SOIL (probably replacing it with libpng plus the DDS loading that we already have) due to it not working as well as we would like. (Unfortunately I don't remember all the specifics about the problems that SOIL has – I don't work on the graphics side of Dolphin after all. But to give you an example of a problem with SOIL, look at the `TODO: remove SOIL usage here` comment in HiresTextures.cpp.) The more formats we support right now, the harder SOIL is going to be to drop without losing functionality.

I don't see how the main feature of this PR would be removed if you don't make Dolphin support additional extensions. You can still count e.g. DDS textures loaded vs PNG textures loaded (and JPG/TGA/BMP too, until/unless SOIL gets dropped).

Anyway, I'm not going to be the person who merges or closes this PR, so that I requested you to remove the support for additional extensions doesn't in itself mean that this PR can't get merged without doing that. Feel free to try to convince someone else to merge the PR with the new extensions intact... But the reason that I requested you to remove them was because I don't think that others are going to want to merge the PR with them intact.

ZexaronS commented:
Whatever then, I just tried not throwing work away, but it appears this is just one example where it won't work. Indeed this PR was made with several mistakes, the current solution going forward doesn't anymore look anything like the first post of this PR.

So, aborting this, the first part of the continuation is here: #5996

spycrab commented:
@dolphin-emu-bot rebuild
